### DISCUSSION TITLE
Query about the Adam optimiser

### DISCUSSION THREAD

**Discussion by BGN3747B** [Part-II-Computational-Physics/.github#126] [Q&A]
I have implemented both the Adam and Lion optimisers using very similar code. However, my Adam optimiser does not seem to converge. 
I've provided my code below. m, v, m_hat, and v_hat have all been initialised as Numpy arrays. 
``` python
        # update (adam)
        lr = eta * learning_rate(step, n_steps)
        #TO DO
        for i in range(Nparam):
            m[i] = b1 * m[i] + (1 - b1) * (model.parameters()[i].grad)
            v[i] = b2 * v[i] + (1 - b2) * (model.parameters()[i].grad)**2
            m_hat[i] = m[i] / (1 - b1**t)
            v_hat[i] = v[i] / (1 - b2**t)
            model.parameters()[i].data = model.parameters()[i].data - lr * m_hat[i] / ((np.sqrt(v_hat[i]) + eps) + lmbda*model.parameters()[i].data)
```

I'm not sure what the problem is. I've tried changing the learning rate, to no avail.

**Comment by AustenLamacraft** [comment]
Would you be able to post all the hyperparameters you are using? That will help in replicating your issue

**Reply by BGN3747B** [reply]
I haven't changed any from what was given:
    def initialise(n_neurons):
    """Function that initialises a 2-layer neural network with n_neruons neurons in each layer."""
    
    model = MLP(2, [n_neurons, n_neurons, 1]) # 2-layer neural network
    print(model)
    print("number of parameters", len(model.parameters()))
    
    return model
model = initialise(n_neurons=16)
And Adam parameters were default values as well:   
    b1 = 0.9
    b2 = 0.999
    eps = 1e-8
    lmbda = 1.0
    eta = 5e-2
    I haven't changed them from the default.

**Reply by BGN3747B** [reply]
I've changed the implementation - now it converges, but it gives a higher accuracy than the Lion optimiser, which seems wrong.
